### DISCUSSION TITLE
Memory Optimization for ItemsRepeater with 600,000 Items in AvaloniaUI 11.1.0 beta 1

### DISCUSSION THREAD

**Discussion by kokaKOLA-NEW** [AvaloniaUI/Avalonia#15600] [Q&A] [answered]
I have implemented an ItemsRepeater in my AvaloniaUI application to display a list of approximately 600,000 items. The current code works, but it consumes around 600 MB of RAM, which is concerning in terms of performance and resource usage.

Here is the code I am currently using:
```axaml 
<ItemsRepeater ItemsSource="{Binding FileResults}">
    <ItemsRepeater.Layout>
        <StackLayout Spacing="5" DisableVirtualization="False"
                     Orientation="Vertical" />
    </ItemsRepeater.Layout>

    <ItemsRepeater.ItemTemplate>
        <DataTemplate>
            <!-- My Item -->
        </DataTemplate>
    </ItemsRepeater.ItemTemplate>
</ItemsRepeater>
```

I am looking for ways to optimize memory usage to efficiently handle this number of items. I am currently using AvaloniaUI version 11.1.0 beta 1.

Any suggestions or best practices to improve ItemsRepeater performance and reduce memory consumption would be greatly appreciated. Thank you in advance for your help!

![image](https://github.com/AvaloniaUI/Avalonia/assets/158836870/fd825213-6893-4255-a787-e9bdc77ddb09)

**Comment by LaurentInSeattle** [comment]
Just started with Avalonia, so I might be wrong.

It seems like the Repeater is not using virtualization by default, or maybe cannot have it. 
You need to use a control that allow virtualization. maybe ItemsControl ?

**Reply by kokaKOLA-NEW** [reply]
You may be right, so I added `DisableVirtualization="False"`  just in case. But it didn't change anything, I'll try with ItemsControl, thanks anyway.

**Reply by thevortexcloud** [reply]
`StackLayout` is virtualised and can't be turned off. To actually disable it you need to use `NonVirtualizingStackLayout`.

**Comment by thevortexcloud** [comment]
1.  What are you displaying? 
2. How many things are visible on screen?
3. How are you adding items to the repeater?
4. How are you cleaning up?

I display ~3k bitmaps in one of my apps using the items repeater and rarely get more than ~200mb-300mb memory usage (when loading images from the cache, resizing them to put into the cache is a different story).  I can get this down to a rock solid ~150mb by just force disposing the bitmaps when they are not needed any more.

**Reply by kokaKOLA-NEW** [reply]
In each item, I display 4 TextBlocks and 1 button. I see 6 items. Here is the part of my code in my function that adds the items, and I clear the ItemSource with `FileResults.Clear()`

```c#
FileResults.Clear();

var progress = new Progress<FileResult>(fileResult => { FileResults.Add(fileResult); });

await Task.Run(() =>
{
    for (var i = 0; i < resultCount; i++)
    {
        long dateModified;
        long size;
        Everything_GetResultDateModified(i, out dateModified);
        Everything_GetResultSize(i, out size);

        var fileName = Marshal.PtrToStringUni(Everything_GetResultFileName(i));
        var filePathBuilder = new StringBuilder(260);
        Everything_GetResultFullPathName(i, filePathBuilder, (uint)filePathBuilder.Capacity);
        var filePath = filePathBuilder.ToString();

        var fileResult = new FileResult
        {
            Name = fileName,
            Path = filePath,
            DateCreated = DateTime.FromFileTime(dateCreated),
            Size = size
        };

        ((IProgress<FileResult>)progress).Report(fileResult);
    }
});
```

**Reply by timunie** [reply]
Hmmm 🤔 4 texrblocks and a button should not load that much memory. That should be fine. Looks like maybe your viewModel leak some memory for fileinfo? Is it disposable?

**Reply by kokaKOLA-NEW** [reply]
Upon reviewing the code, I did not find any obvious memory leaks related to the `FileInfo` or `FileResult` objects. I will continue to explore potential optimization opportunities to further improve memory usage and overall performance.

**Reply by kokaKOLA-NEW** [reply]
I just upgraded to the latest stable version of AvaloniaUI and ItemsRepeater, and it significantly reduced memory usage.

<img width="143" alt="image" src="https://github.com/AvaloniaUI/Avalonia/assets/158836870/da2c86eb-b3ed-4bb9-a016-1823561282cf">

**Reply by thevortexcloud** [reply, marked as answer]
You seem to be using Marshal and native pointers. Without seeing the rest of your code it's possible you may have a memory leak, if you do not ever free the native memory allocated to the pointers region.

It also looks like you are eager loading your data, as such based on some quick math and some assumptions:

* Assuming each file name is ~20 characters, that would be:  20 bytes
* Assuming each path is 260 characters, that would be 260 bytes
* Assuming a DateTime takes up ~100 bytes (I have no idea what the actual size of it is so that's a complete guess)
* A long takes up 64 bits. or 8 bytes

Which means you are looking at roughly 388 bytes per object. Multiply that by 600000 you get 232 megabytes of memory being used, just from your collection. Obviously the real numbers are probably going to be different and possibly much higher as strings are variable length and the size of them can vary by the text encoding. It's entirely possible here that for the most part, the problem has nothing to do with Avalonia and is more to do with how much data you are keeping in memory. If it's possible to lazy load some of this instead you will likely be able to save a lot of memory. Alternatively, pagination can also work well but be real annoying to implement.

Have you also run a memory profile tool to see what is actually using the memory? There is one built into Visual Studio and Rider. It would tell you where you should focus your attention.

**Reply by thevortexcloud** [reply]
> I just upgraded to the latest stable version of AvaloniaUI and ItemsRepeater, and it significantly reduced memory usage.

Excellent. I still think there is room to possibly optimise your code, but if you are happy with that then I would say you can close this discussion.

**Reply by kokaKOLA-NEW** [reply]
Thank you very much.
